src/functions/includegraphics.js

```diff
@@ -0,0 +1,167 @@
+// @flow
+import defineFunction from "../defineFunction";
+import type {Measurement} from "../units";
+import {calculateSize, validUnit} from "../units";
+import ParseError from "../ParseError";
+import {Img} from "../domTree";
+import mathMLTree from "../mathMLTree";
+import {assertNodeType, checkNodeType} from "../parseNode";
+
+import type {AnyParseNode} from "../parseNode";
+import type {CssStyle} from "../domTree";
+
+// TODO: make it easier for commands to get the macro expanded text
+// from inside an ordgroup argument.
+const stringFromParseGroup = function(textArray: AnyParseNode[]): string {
+    let str = "";
+    for (let i = 0; i < textArray.length; i++) {
+        const textord = checkNodeType(textArray[i], "textord");
+        const spacing = checkNodeType(textArray[i], "spacing");
+        if (textord) {
+            str += textord.text;
+        } else if (spacing) {
+            str += spacing.text;
+        }
+    }
+    return str;
+};
+
+const sizeData = function(str: string): Measurement {
+    if (/^[-+]? *(\d+(\.\d*)?|\.\d+)$/.test(str)) {
+        // str is a number with no unit specified.
+        // default unit is bp, per graphix package.
+        return {number: +str, unit: "bp"};
+    } else {
+        const match = (/([-+]?) *(\d+(?:\.\d*)?|\.\d+) *([a-z]{2})/).exec(str);
+        if (!match) {
+            throw new ParseError("Invalid size: '" + str
+                + "' in \\includegraphics");
+        }
+        const data = {
+            number: +(match[1] + match[2]), // sign + magnitude, cast to number
+            unit: match[3],
+        };
+        if (!validUnit(data)) {
+            throw new ParseError("Invalid unit: '" + data.unit
+             + "' in \\includegraphics.");
+        }
+        return data;
+    }
+};
+
+defineFunction({
+    type: "includegraphics",
+    names: ["\\includegraphics"],
+    props: {
+        numArgs: 1,
+        numOptionalArgs: 1,
+        argTypes: ["text", "text"],
```

Comment:
When I first wrote this, I tried to specify a `url` arg type but got an error. I'll try again and see if I can identify the problem.

Reply:
@ronkok The `url` argument is hardcoded in the Lexer and the Parser:

https://github.com/Khan/KaTeX/blob/f71f469d4b3ca97654702a0807e736d176f4c4c5/src/Lexer.js#L49-L52

https://github.com/Khan/KaTeX/blob/9b7e10e4f753786f125254a79d7cd88acc08afd4/src/Parser.js#L916-L917

You'll need to update regexes.

Reply:
Regarding a `url` arg type. There really isn't any such thing in KaTeX. I've experimented a bit with `href.js` and replaced `argTypes: ["url"]` with `argTypes: ["text"]` and `argTypes: ["chaos"]`. They both run fine. It doesn't matter what arg type is specified in `href.js` because `Parser.js` ignores that specification.

What is really happening is that `Lexer.js` is preempting the usual logic for functions `\href` and `\url`. That `urlFunctionRegexString` captures both the function name and the first argument group. `Parser.js` then handles the result as an exception to the regular run of parsing code.

That RegEx, by the way, is not detailed enough to ensure that the argument is a [valid url](https://stackoverflow.com/questions/161738/what-is-the-best-regular-expression-to-check-if-a-string-is-a-valid-url#). It seems to be just enough to capture the function name and its first argument group.

So I question the usefulness of changing the code in this PR. The `\href` example lexes some arbitrary text then parses it as an exception to regular input. This PR does the same thing; it just does it in a different file. And, I would argue, by placing this exception code in the `includegraphics.js` file, I am grouping exeption code together in manner similar to the way we currently use `defineFunction`.

Reply:
@ronkok In `Parser`, there are escaping routines and allowed protocol checks:
https://github.com/Khan/KaTeX/blob/9b7e10e4f753786f125254a79d7cd88acc08afd4/src/Parser.js#L926-L938

Reply:
@ylemkimon  That's an excellent point. I'll move that code into a utility function so that it is available to `includegraphics`. 

If someone wants to refactor Lexer and Parser in an way that creates an actual `url` arg type, that would be great, but that's not for me.

Reply:
Sounds good.

> argTypes: ["chaos"]. They both run fine. 

That feels like a bug.  @ronkok could you file an issue for this?